### PR TITLE
[no squash] NetworkPacket and other improvements

### PR DESCRIPTION
**Pull request by sfan5**
commit 1: fixes https://irc.minetest.net/minetest-dev/2024-01-24#i_6148094 and prevents it from happening again
commit 2: it has happened in the past that someone calculated the size passed to the `NetworkPacket` ctor wrong which would leave unused null bytes at the end. reserving still has the speed advantage but is safer.
the rest: I got distracted and found more things to improve.

## To do

This PR is Ready for Review.

### CODE REVIEW

**Comment by lhofhansl** [comment]
Looks sensible. I find it hard to feel sure that this does not break anything.
Anything specific to test?

**Comment by sfan5** [comment]
For full test coverage you'd test every code path in the server (or client) that generates a packet until you have covered all types.
In practice if you compare networkprotocol.h and check that all `TOCLIENT_*` and `TOSERVER_*` are also contained in the relevant *opcodes.cpp file everything should be fine.

**Review by SmallJoker — APPROVED** [review]
:ballot_box_with_check:  Joined an older server
:ballot_box_with_check: checked whether the TOCLIEnT_ and TOSERVER_ messages are indeed unused since 5.0.0 (start of network compat).

Works for me.

File: src/server.cpp [hunk 1]
```diff
@@ -2137,9 +2113,8 @@ void Server::SendActiveObjectMessages(session_t peer_id, const std::string &data
 
 	pkt.putRawString(datas.c_str(), datas.size());
 
-	m_clients.send(pkt.getPeerId(),
-			reliable ? clientCommandFactoryTable[pkt.getCommand()].channel : 1,
-			&pkt, reliable);
+	auto &ccf = clientCommandFactoryTable[pkt.getCommand()];
```
[the remark this answers is not in the document]
Comment (sfan5) [reply src/server.cpp:2116]: `sendCustom` already does

**Pull request merged**